### PR TITLE
Feature: filter points of interest by opening times

### PR DESCRIPTION
**Pull request by digorath**
- implemented explicit closed times for POIs
- added a filter for showing only currently open POIs on the category level

---

SVA-319

## Screenshot

Simulator Screen Shot - iPhone 12 Pro - 2021-11-02 at 10 13 50

### CODE REVIEW

**Review by donni106 — CHANGES_REQUESTED** [review]
huh. hard one. there are very complex methods i do not have easier logics for. but this can result in struggle in the future. maybe we can separate some stuff from momentHelper to an own helper? there is a lot of stuff that is not "just using moment method" outsourced here.

File: src/helpers/momentHelper.js [hunk 1]
```diff
@@ -137,6 +137,196 @@ const getTodayWithTime = (time) => {
...
+  const sortedIntervals = [...intervals].sort((a, b) => {
+    if (!a.timeFrom && !b.timeFrom) {
+      return 0;
+    }
+    if (!a.timeFrom) {
+      return -1;
+    }
+    if (!b.timeFrom) {
+      return 1;
+    }
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:229]: i do not quite get what you are saying.

> why is an undefined timeFrom of a considered "smaller" than an undefined timeFrom of b?
it is not. they are treated as equal in the first if block.

> shouldn't the undefined ones be sorted at the end or beginning in any case?
they are, are they not? 🤔

File: src/helpers/momentHelper.js [hunk 2]
```diff
@@ -137,6 +137,196 @@ const getTodayWithTime = (time) => {
...
+    return a.timeFrom < b.timeFrom ? -1 : 1;
+  });
+
+  // merge intervals starting from the one that starts first
+  return sortedIntervals.reduce(
+    (mergedIntervals, currentInterval) => {
+      const latest = mergedIntervals.pop();
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:236]: the most recent one is the last one that was pushed onto the `mergedIntervals´. why do you think it should be the first one?

it is also the latest in terms of "hour of the day" of the entries in `mergedIntervals`.

File: src/helpers/momentHelper.js [hunk 3]
```diff
@@ -137,6 +137,196 @@ const getTodayWithTime = (time) => {
...
+
+      // if they overlap, use the merged interval
+      if (result) {
+        mergedIntervals.push(result);
+        return mergedIntervals;
+      }
+
+      // if they do not overlap, then the latest one will not overlap with any others, and is completely merged already.
+      // in that case the current interval needs to be checked next
+      mergedIntervals.push(latest, currentInterval);
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:247]: > because of the pop in the next round this check means to be the next "latest" in reduce step?

correct. i should rephrase it to "needs to be checked in the next step of reducing the array"

File: src/helpers/momentHelper.js [hunk 4]
```diff
@@ -137,6 +137,196 @@ const getTodayWithTime = (time) => {
...
+// eslint-disable-next-line complexity
+const mergeIntervals = (timeA, timeB) => {
+  // check which one starts first
+  let isAFirst = false;
+
+  if (timeA.timeFrom && timeB.timeFrom) {
+    isAFirst = timeA.timeFrom < timeB.timeFrom;
+  } else if (!timeA.timeFrom) {
+    isAFirst = true;
+  }
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:182]: that would be equivalent yes. i will change it.

File: src/helpers/momentHelper.js [hunk 5]
```diff
@@ -137,6 +137,196 @@ const getTodayWithTime = (time) => {
...
+  const first = isAFirst ? timeA : timeB;
+  const second = isAFirst ? timeB : timeA;
+
+  // check if the one that starts first ends after the other one started, which means that they do not overlap
+  if (first.timeTo && second.timeFrom && first.timeTo < second.timeFrom) {
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:187]: we would need something like `if(first.timeTo && moment(first.timeTo).isBefore(second.timeFrom))`.

we could not omit the check if `first.timeTo` is present because `moment(undefined)` will evaluate to the current time.

i do not see that as being more readable, and neither do i think it would make a comment less necessary.

File: src/helpers/momentHelper.js [hunk 6]
```diff
@@ -137,6 +137,196 @@ const getTodayWithTime = (time) => {
...
+    // if this is the case, they do not overlap
+    return;
+  }
+
+  // find the later ending time and return with the start time of the first one
+  // if one of them does not end, the merged one does not end
+  if (!first.timeTo || !second.timeTo) {
+    return { timeFrom: first.timeFrom };
+  }
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:196]: it is a description of the remaining body of the function. the `timeTo` being `undefined` means that it is open ended and should be treated as the "later ending time", so i see this as applying to both the `if`-block and the final return.

File: src/helpers/momentHelper.js [hunk 7]
```diff
@@ -137,6 +137,196 @@ const getTodayWithTime = (time) => {
...
+    return;
+  }
+
+  const closed = closedIntervals.find((interval) =>
+    dateIsWithinInterval(time, interval.timeFrom, interval.timeTo)
+  );
+
+  if (closed) {
+    if (!closed.timeTo) {
+      return undefined;
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:282]: hmmm. good question. i will remove it for consistency.

File: src/helpers/momentHelper.js [hunk 8]
```diff
@@ -137,6 +137,196 @@ const getTodayWithTime = (time) => {
...
+const findNextOpenTime = (openIntervals, closedIntervals, time) => {
+  let opens = openIntervals.filter((interval) => !interval.timeTo || interval.timeTo > time);
+
+  if (!opens.length) {
+    return;
+  }
+
+  const closed = closedIntervals.find((interval) =>
+    dateIsWithinInterval(time, interval.timeFrom, interval.timeTo)
+  );
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:278]: the next opening times could overlap with closing times, so we need to check for closing times as well, as they are of higher priority than opening times.

File: src/helpers/momentHelper.js [hunk 9]
```diff
@@ -161,59 +351,63 @@ export const isOpen = (openingHours) => {
...
-      currentlyValidOpeningTimes.forEach(({ timeTo }) => {
-        timeDiff = Math.max(timeDiff, moment(timeTo).diff(now, 'minute'));
-      });
+    // check if there is a current interval of being explicitly closed or open
+    const currentClosedInterval = sortedClosedIntervals.find((interval) => {
+      return dateIsWithinInterval(now, interval.timeFrom, interval.timeTo);
+    });
+    const currentOpenInterval = sortedOpenIntervals.find((interval) => {
+      return dateIsWithinInterval(now, interval.timeFrom, interval.timeTo);
+    });
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:382]: no particular reason i think. i resorted the code a lot while developing, this is just an artifact of that i think. i will change it here to be in line with other code.

File: src/queries/pointsOfInterest.js [hunk 10]
```diff
@@ -61,6 +61,17 @@ export const GET_POINTS_OF_INTEREST = gql`
...
+        id
+        sortNumber
+        weekday
+        timeFrom
+        timeTo
+        open
+        dateFrom
+        dateTo
+        description
+      }
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/queries/pointsOfInterest.js:74]: i think neither the `sortNumber` nor the `description` are needed. `id` is in this query as always to avoid caching problems.

i will remove the two unused fields from this query.

File: src/screens/IndexScreen.js [hunk 11]
```diff
@@ -91,7 +98,7 @@ export const IndexScreen = ({ navigation, route }) => {
 
   // if we show the map or want to sort by distance, we need to fetch all the entries at once
   // this is not a big issue if we want to sort by distance, because getting the location usually takes longer than fetching all entries
-  if (showMap || sortByDistance) {
+  if (showMap || sortByDistance || filterByOpeningTimes) {
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/screens/IndexScreen.js:101]: will do ✔️

File: src/screens/IndexScreen.js [hunk 12]
```diff
@@ -149,7 +156,16 @@ export const IndexScreen = ({ navigation, route }) => {
...
     <SafeAreaViewFlex>
       {query === QUERY_TYPES.POINTS_OF_INTEREST ? (
-        <MapSwitchHeader setShowMap={setShowMap} showMap={showMap} />
+        <View>
+          <MapSwitchHeader setShowMap={setShowMap} showMap={showMap} />
+          <WrapperHorizontal>
+            <WrapperRow style={styles.textWrapper}>
+              <RegularText>{texts.pointOfInterest.filterByOpeningTime}</RegularText>
+              <Switch switchValue={filterByOpeningTimes} toggleSwitch={setFilterByOpeningTimes} />
+            </WrapperRow>
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/screens/IndexScreen.js:165]: i will check if there is a reason not to use it as in the waste calendar reminder.

i don't like putting an `ActivityIndicator` there, as we would need to somehow link it to the loading state of the query, and then it would be spinning for any initial loading as well. and arbitrary spinner shown for a fixed period also seems wrong to me. the switching is not the thing that is in a loading state here anyways.

do you have a better idea?

File: src/screens/IndexScreen.js [hunk 13]
```diff
@@ -237,3 +259,10 @@ IndexScreen.propTypes = {
   navigation: PropTypes.object.isRequired,
   route: PropTypes.object.isRequired
 };
+
+const styles = StyleSheet.create({
+  textWrapper: {
+    alignItems: 'center',
+    justifyContent: 'space-between'
+  }
+});
```
[the remark this answers is not in the document]
Comment (digorath) [reply src/screens/IndexScreen.js:268]: ✔️

File: src/helpers/momentHelper.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js:247]: rephrased the comment now.
would renaming `latest` to `previous` change anything for you? do you have other suggestions?

**Comment by digorath** [comment]
> maybe we can separate some stuff from momentHelper to an own helper?

we can extract most of the logic into an `openingTimesHelper` if you want. it is correct, that much of it is not using moment.js.
i used the momentHelper file more as a "time manipulation" helper, but even then a lot of the logic is specific to our format of opening times

File: src/screens/IndexScreen.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (donni106) [reply src/screens/IndexScreen.js]: if there is no usable loading state then we can not have one.

File: src/helpers/momentHelper.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (donni106) [reply src/helpers/momentHelper.js:186]: ok, using moment in this case is different but not better. we keep it as is.

File: src/helpers/momentHelper.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (donni106) [reply src/helpers/momentHelper.js:195]: got it

File: src/helpers/momentHelper.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (donni106) [reply src/helpers/momentHelper.js:228]: ah jo, my fault was to think in a and b as different cases but we need to think of one case comparing a and b, thus we need to return either -1 or 1.

File: src/helpers/momentHelper.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (donni106) [reply src/helpers/momentHelper.js:246]: good idea, i would bring it even further to `previousInterval` which makes much sense using together with `currentInterval`

File: src/helpers/momentHelper.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (donni106) [reply src/helpers/momentHelper.js:235]: could be resolved with https://github.com/ikuseiGmbH/smart-village-app-app/pull/385/files/c6e5d0ac9ce3ee9f3a573160c0f3db90588b0697#r748367992

File: src/helpers/momentHelper.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (donni106) [reply src/helpers/momentHelper.js:277]: got it

[the remark this answers is not in the document]
Comment (donni106) [reply src/helpers/momentHelper.js:277]: would be nice to see some tests where we could add random scenarios to see if different cases succeed. i will think of some example intervals as well.

File: src/helpers/momentHelper.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js]: ✅

File: src/helpers/momentHelper.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js]: ✅

File: src/helpers/momentHelper.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (digorath) [reply src/helpers/momentHelper.js]: added tests 💯

File: __tests__/helpers/openingHoursHelper.test.js [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+const FAKE_NOW_DATE = new Date();
+FAKE_NOW_DATE.setHours(12);
+FAKE_NOW_DATE.setMinutes(0);
+FAKE_NOW_DATE.setSeconds(0);
+FAKE_NOW_DATE.setMilliseconds(0);
+
+const isOpenWithFakeTimeDateTime = (openingHours) => isOpen(openingHours, FAKE_NOW_DATE);
+
+describe('testing the opening times handling', () => {
+  test('single opening intervals work as expected', () => {
```
Comment (donni106) [inline __tests__/helpers/openingHoursHelper.test.js:14]: how come that you are using `describe/test` instead of `describe/it` like we have in the rest of our tests?

File: src/helpers/openingHoursHelper.ts [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+ * WARNING: testDate is only for tests, and will yield unexpected result if the day is not the present day.
+ *
+ * Checks for a given array of opening times, if it is currently open, and if it opens or closes soon.
+ * The timeDiff contains the remaining minutes until it opens or closes,
+ * timeDiff may be undefined, if there are opening times where either timeFrom or timeTo is missing.
+ */
+// eslint-disable-next-line complexity
+export const isOpen = (
+  openingHours: OpeningHour[],
+  testDate?: Date
```
Comment (donni106) [inline src/helpers/openingHoursHelper.ts:225]: i have problems with this "test-only" parameter we need to take care of and describe in comments. what about making the `now` our parameter here with `new Date()` as default value? when calling this method in tests, we could pass different values. this way we have no specific case we need to warn about and think of, right? we can "just use it" the way we need the method, right?

**Review by donni106 — APPROVED** [review]
thanks for changes. looking good. i will now check the test scenarios more in detail and think about adding further cases.

File: __tests__/helpers/openingHoursHelper.test.js [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+    expect(
+      isOpenWithFakeTimeDateTime([
+        {
+          open: true,
+          timeFrom: '10:00',
+          timeTo: '17:00',
+          weekday: todayString
+        }
+      ])
+    ).toEqual({ open: true, timeDiff: 300 });
```
Comment (donni106) [inline __tests__/helpers/openingHoursHelper.test.js:34]: this first test block contains just "....open" checks except of this one. wouldn't it be better to also check .open equals true here and ignore the `timeDiff` with reason?

File: __tests__/helpers/openingHoursHelper.test.js [hunk 17]
```diff
@@ -0,0 +1,215 @@
+import { getReadableDay, isOpen } from '../../src/helpers';
+
+const todayString = getReadableDay(new Date());
+
+const FAKE_NOW_DATE = new Date();
```
Comment (donni106) [inline __tests__/helpers/openingHoursHelper.test.js:5]: i think it would be more clear to get the weekday from the time that is checked in the methods, in this case `const weekday = getReadableDay(FAKE_NOW_DATE)`. i know at the end it will result in the same, but for readability and context, you know? and with using `weekday` variable we could get rid of every `todayString` in jsons and just use `weekday`, right?!

File: src/helpers/openingHoursHelper.ts (hunk 15 above)
Comment (digorath) [reply src/helpers/openingHoursHelper.ts]: i changed `testDate` to now with fallback of `new Date()`,  and changed the warning. ( the date passed still needs to be "today" because the behaviour will not be as expected otherwise.

**Review by digorath — COMMENTED** [review]
added changes after review

**Review by donni106 — APPROVED** [review]
